docs/config.md
# 設定値

koyomiには次のプロパティと既定値が設定されています  
プロパティを変更することで、実務に沿った暦に設定します  

## フォーマット

フォーマットの既定値は`'YYYY-MM-DD HH:II:SS'`です  
この値を変更した場合は、`format`メソッドで第二引数を指定しなかった場合に既定値が使用されます  
設定するには次のようにします  

```
koyomi.defaultFormat = 'Y年M月D日 H時I分S秒';
```

フォーマットの書式に関しては[フォーマット ./format.md](./format.md)を参照してください

## 定休日

定休日は週や月で決められた休業日のことです  
既定値は`'土,日'`です  
複数指定するときには、カンマ区切りで指定します  
`'sun'`や`'Sunday'`なども設定することができます  
固定の日を指定する場合は、`'5,15,25'`のように数字を指定します  
第2・3水曜日などの指定は`'2水,3水'`と、何週目のあとに曜日を指定します  
5以上を指定すると最終週の指定とみなします  
これらは、混合させることもできます  
上記では対応できない場合は、関数を指定することで細かな定休日を指定することができます  
引数はDateオブジェクトのみで、定休日の場合はtrueを返すようにします  
休業の事由は`'定休日'`です

```
koyomi.regularHoliday = '日,2土,3土,30';
```

## 年末年始、お盆の休業日

定休日とは別に、年末年始・お盆の休業日を設定することができます  
こちらは休業の事由とともに日付もしくは日付の期間を設定します  
既定値は、`'年末年始のお休み 12/29-1/3'`です  
事由は省略することできますが、その場合は`'休業期間'`が事由に設定されます  
複数の期間も指定する場合は、カンマ区切りで指定します  
単発の日付を指定する場合は、`'6/23, 10/10'`などとします  
上記では対応できない場合は、関数を指定することで細かな休業日を指定することができます  
引数はDateオブジェクトのみで、休業日の場合は休業の事由を文字列で返すようにします  
事由は`koyomi.closeCause`やカレンダー情報取得で使用します  

```
koyomi.seasonHoliday = '年末年始のお休み12/29-1/3, お盆のお休み8/16-8/18, 創立記念日10/10';
```

## 祝日の営業

祝日を営業日にするかどうかを設定することができます  
既定値は`false`(祝日は休業日)です  
休業の事由には祝日名が返されます

```
koyomi.openOnHoliday = true; // 営業日にする
```

## 年の始まり

一年の始まりを設定することができます  
既定値は`1`(1月)です  
年度の範囲を設定する重要なプロパティで様々なメソッドに影響します  
標準的な会計年度にあわせ4月に設定する場合は次のようにします  

```
koyomi.startMonth = 4;
```

## 週の始まり

週の始まりを設定することができます  
既定値は`'日'`です  
週の開始・終了・期間を計算する場合やカレンダーデータの作成などで使用します  
月曜日へ変更する場合は、次のようにします  

```
koyomi.startWeek = '月';
```

# 暦オブジェクト

koyomiはクラスではありません  
そのため、`new`を使用したインスタンスの作成はできません  
その代わりに、新規作成や複製のメソッドで新しい暦オブジェクトを取得することができます  

## 新規作成

createメソッドで新規作成してください  
既定値の設定の暦オブジェクトを取得することができます

```
var created = koyomi.create(); 
```

## 複製

cloneメソッドで複製してください  
同じ設定値で複製されますが、それぞれが独立して設定値の変更することができます

```
var copied = koyomi.clone(); 
```


# ドキュメント一覧

  + [イントロダクション ../README.md](../README.md)
  + 設定値 ./config.md
  + [フォーマット ./format.md](./format.md)
  + [日時の情報取得・操作 ./calc-date.md](./calc-date.md)
  + [営業日計算 ./calc-biz.md](./calc-biz.md)
  + [カレンダー情報 ./calendar.md](./calendar.md)
  + [祝日 ./holiday.md](./holiday.md)
  + [営業・休業 ./open-close.md](./open-close.md)
  + [補助関数 ./helper.md](./helper.md)

Changelog.md
# Change Log

 0.5.2 / 2015-12-08
===================
  
  * テストのutil/create.js、util/clone.jsを記述
  * example/testを追加
  * docs/config.md、startMonthの既定値を修正


 0.5.1 / 2015-12-06
===================

  * v1.0に向け大幅に整理しました
  * **注意! v0.3.9までのAPIと全く互換性がありません**
  * ES6(ES2015)で記述
    * クライアントサイドはbabelによりes5のコードに変換済みです
  * イントロダクション＆インスタンスの作成
    * グローバル変数Koyomiをkoyomiに変更。クラスではなくオブジェクトに変更しました
    * クラスメソッドとインスタンスメソッドの区別がなくなりすべて共通のオブジェクトの関数になりました
    * インスタンスの作成が不要になったため、ドキュメントの"インスタンスの作成"は"設定値"と名称変更しました
    * 年が関連するメソッドやパラメータの1月始まりを廃止し、startMonthに統一しました
    * 設定値holidayOpenedはopenOnHolidayに変更になりました
    * seasonHolidayに休業の事由を設定できるようになりました
    * koyomi.createとkoyomi.cloneメソッドが追加されました
  * フォーマット
    * パラメータ文字列を整理しました
        * VV, Vを廃止し、VxをVに変更
        * Qを廃止し、QxをQに変更
        * R,Rjを廃止し、Rx,RiをR,rに変更、
        * TT,T,tを廃止し、TTx,Tx,txをTT,T,tに変更
        * Wをwに、WjをWに変更
        * FFwをFFに、FFをFに、Fwをffに、Fをfに変更
        * EwをEEに変更
        * eの戻り値に特別休業と祝日名を返すよう変更
        * Aをaaに、AjをAに変更
    * formatの引数をコンパイルしキャッシュするように変更し高速化しました
    * formatNendoをformatYearに名称変更し、ドキュメントをフォーマットの項に移動させました
    * formatYearの引数はformatNendoと異なります
  * 日時の情報取得・操作
    * 新しいメソッドgetRangeが追加されました
        * 指定した単位の範囲で最初と最後の日時を返します
    * start,endをfrom,toに名称変更しました
    * diffDays,diffMinutes,diffSecondsを、汎用化しdiffに統合しました
    * yearDays,monthDaysを汎用化しdaysに統合しました
    * passYearDays,passDaysを汎用化しpassDaysに統合しました
    * remainYearDays,remainDaysを汎用化しremainDaysに統合しました
    * 新しいメソッドseparateが追加されました
        * 期間をブロックごとに分割(年度、月、日)されます
    * kindのソースコードをcalcDate.jsに移動しました
  * 営業日計算
    * コードを記述していたファイルがcalcEigyobi.jsからcalcBiz.jsに変更になりました
    * addEigyobi,toEigyobiをaddBiz,toBizに名称変更しました
    * countEigyobi,nendoEIgyobi,monthEigyobiは汎用化しbizに統合しました
    * passNendoEigyobi,passMonthEigyobiは汎用化しpassBizに統合しました
    * remainNendoEigyobi,remainEigyobiは汎用化しremainBizに統合しました
  * カレンダー情報
    * getCalendarDataの引数はrangeだけになりました
    * 引数でしていたstartWeekはオブジェクトの設定値を常に使用します
    * 引数やインスタンスで指定していた設定値sixを廃止し、カレンダーデータは常に6週分返すようにしました
    * ゴースト日が不要な場合は、手動で戻り値をArray.filterで排除してください
    * 各日付のプロパティopened,closedはopen,closeに名称を変更しました
    * 週番号の取得getWeekNumberのドキュメントが補助関数を重複していたので、削除しました
    * 日のデータの取得を効率化しました（formatの代わりにformatArrayを使用）
  * 年度
    * 年度のドキュメントは削除されました
        * 1月始まりではなくstartMonthに統一されたため
    * getNendoはgetEraとして補助関数の項目に移動しました
    * nendoDays,passNendoDays,remainNendoDaysは廃止され、それぞれdays,passDays,remainDaysに統合されました
    * formatNendoはformatYearとしてフォーマットのドキュメントに移動しています
  * 祝日
    * 特に変更はありません
  * 営業・休業
    * isOpenedはisOpenに名称変更しました
    * isCloseが追加されました
    * forcedOpen,forcedCloseはisSetOpen,isSetCloseに名称変更しました
    * isHolidayClosedはisHolidayCloseに名称変更しました
  * 補助関数
    * フォルダ名がfxからutilsに変更になりました
    * utilsに配置された関数はすべて単体で動作するようになりました
        * thisに依存しません
    * parseメソッド追加になりました
    * toDateはtoDatetimeに名称が変更になり、第二引数がなくなりました
    * toDateが新しく日付のみを返す関数として追加されました。旧toDate(value, ture)です
    * toDate,toDatetimeは、文字列・数字を引数にした場合に存在する日時を指定しなければnullを返すようになりました
    * toDate,toDatetimeの月が省略された場合は、startMonthが設定されるようになりました
    * getDateArray,getWeekIndexは廃止されました
    * getNengoはgetEraに名称変更しました
  * その他
    * const.jsはconfig.jsに統合されました
    * Array.indexOfを使用しているコードのいくつかをmapを使用するように変更し、高速化しました

 0.3.9 / 2015-09-10
===================

  * パラメータKを追加
  * kindを追加


  0.3.8 / 2015-09-04
===================

  * パラメータOを追加
  * getAgeを追加


 0.3.7 / 2015-08-25
===================

  * パラメータ文字列のYYYYを追加
  * formatの規定値をYYYY-MM-DD HH:mm:ssに変更


 0.3.5 / 2015-05-26
===================

  * suji,formatの序数を文字列に対応
  * toDateの文字列時のpm,Pm,午後を含む場合に対応
  * パラメータ文字列のCをeに移動(Cを空ける)


 0.3.4 / 2015-05-20
===================

  * getISOWeekNumberの追加
  * パラメータ文字列 GGG Ri 追加 Mj 削除
  * パラメータ文字列
  * const整理
  * format.jsの定数設定を切り出し、defineMushにパラメータ文字列を後から変更した時に対応可能にする
  * テストにWAREKI,wareki追加
  * forcedOpenClose, forcedOpen, forcedCloseの修正とテスト追加
  * example/calc追加


 0.3.3 / 2015-05-17
===================

  * インスタンスのformatから引数eigyobi,includeを削除


 0.3.2 / 2015-05-15
===================

  * calcDateの年区切り、週初めをconfigから取得
  * テストコードをフォルダに分類
  * パラメータ文字列変更 YYYY->Y, YY->y
  * パラメータ文字列削除 Zx
  * checkDayInfoHoliday -> forcedOpenClose
  * getTani追加


 0.3.1 / 2015-05-14
===================

  * テストコード作成
  * パラメータ文字列変更


 0.3.0 / 2015-04-09
===================

  * ファイル分割


 0.2.1 / 2015-04-06
===================

  * 初期リリース

lib/parameters.js
/**
 * dependencies
 */
var CONFIG = require('./config');

/**
 * *********************************************************
 *               パラメータ文字列の変換定義
 * *********************************************************
 *
 *  ２つの引数をとる関数を設定します
 *    t: Dateのインスタンス
 *    k: koyomiオブジェクト
 *  戻り値が、数値や文字列の場合はその値がパラメーター文字列と置換されます
 *
 *  新たなパラメータ文字列の組み合わせを指定してパラメータ文字列を作成することも
 *  できます。その場合は関数ではなく文字列を指定してください
 *  参考: WAREKI, wareki
 *
 *  また引数の日時をなんらかの計算した後に、変換を行う場合は
 *  配列を戻り値にする関数を設定します
 *  配列では、最初の要素にDateのインスタンスを、
 *  ２番目の要素に新しいパラメータ文字列の組み合わせを指定します
 *  参考: BIZ3
 *
 */
var PARAMS;

module.exports = PARAMS = {

  //  -----  カスタマイズ  -----

  // 和暦-年区切り   平成元年一月七日
  WAREKI: 'GGN年M月D日>>漢数字',

  // 和暦-日付区切り 昭和六十四年一月七日
  wareki: 'ggn年M月D日>>漢数字',

  // ISO 8601 拡張形式
  ISO: '{YYYY}-{MM}-{DD}T{HH}:{II}:{SS}{Z}',

  // 3営業日後の日付を返す
  BIZ3: (t, k) => [k.addBiz(t, 3), 'YYYY-MM-DD'],

  //  -----  西暦  -----

  // 西暦-4桁    2014
  YYYY: t => ('0000' + t.getFullYear()).slice(-4),

  // 西暦 2014
  Y: t => t.getFullYear(),

  // 西暦-下2桁  14
  y: t => ('' + t.getFullYear()).slice(-2),


  // -----  皇紀 -----

  // 皇紀 2675
  GGG: t => t.getFullYear() + 660,


  // -----  和暦 (年区切り)-----

  // 和暦-年号   平成
  GG: (t, k) => k.getEra(t).N,

  // 和暦-年号   H
  G: (t, k) => k.getEra(t).n,

  // 和暦-年     1,  27
  N: (t, k) => t.getFullYear() - k.getEra(t).y + 1,


  // -----  和暦 (日付区切り) -----

  // 和暦-年号  平成
  gg: (t, k) => k.getEra(t, true).N,

  // 和暦-年号  H
  g: (t, k) => k.getEra(t, true).n,

  // 和暦-年    1,  24
  n: (t, k) => t.getFullYear() - k.getEra(t, true).y + 1,


  // ----- 上半期・下半期 -----

  // 上半期, 下半期
  V: (t, k) => (t.getMonth() - k.startMonth + 13) % 12 < 6 ? '上半期' : '下半期',

  // ----- 四半期 -----

  // 1, 2, 3, 4      開始月に依存する四半期
  Q: (t, k) => {
    // startMonthを仮想1月とした時何月か
    var index = (t.getMonth() - k.startMonth + 14) % 12 || 12;
    return parseInt((index - 1) / 3, 10) + 1;
  },


  // -----   月   -----

  // 英語    August
  MMM: t => CONFIG.MONTH[t.getMonth()],

  // 0埋め    08,  12
  MM: t => ('0' + (t.getMonth() + 1)).slice(-2),

  // 8,  12
  M: t => t.getMonth() + 1,


  // -----  週番号  -----

  // 週番号
  R: (t, k) => k.getWeekNumber(t),

  // ISO週番号
  r: (t, k) => k.getISOWeekNumber(t),


  // -----  日  -----

  // 05,  25
  DD: t => ('0' + t.getDate()).slice(-2),

  //  5,  25
  D: t => t.getDate(),


  //  -----  総日数・経過日数・残日数(年)  -----

  // 総日数(年)
  CCC: (t, k) => k.days(t, 'year'),

  // 経過日数(年)
  CC: (t, k) => k.passDays(t, 'year'),

  // 残日数(年)
  C: (t, k) => k.remainDays(t, 'year'),


  //  -----  総日数・経過日数・残日数(月)  -----

  // 総日数(月)
  ccc: (t, k) => k.days(t, 'month'),

  // 経過日数(月)
  cc: (t, k) => k.passDays(t, 'month'),

  // 残日数(月)
  c: (t, k) => k.remainDays(t, 'month'),


  //  -----  営業日数・営業残日数(年)  -----

  // 総営業日数(年)
  BBB: (t, k) => k.biz(t, 'year'),

  // 経過営業日数(年)
  BB: (t, k) => k.passBiz(t, 'year'),

  // 残営業日数(年)
  B: (t, k) => k.remainBiz(t, 'year'),


  //  -----  営業日数・営業残日数(月)  -----

  // 総営業日数(月)
  bbb: (t, k) => k.biz(t, 'month'),

  // 経過営業日数(月)
  bb: (t, k) => k.passBiz(t, 'month'),

  // 残営業日数(月)
  b: (t, k) => k.remainBiz(t, 'month'),


  // -----  曜日  -----

  // Monday
  WW: t => CONFIG.WEEK[t.getDay()],

  // 日本語 日曜日, 月曜日,
  W: t => CONFIG.JWEEK[t.getDay()] + '曜日',

  // 日:0 -> 土:6
  w: t => t.getDay(),


  //  -----  祝日  -----

  // 祝日名または曜日  元日, 成人の日, 日曜日...土曜日
  FF: (t, k) => k.getHolidayName(t) || PARAMS.W(t, k),

  // 祝日名            元日, 成人の日,
  F : (t, k) => k.getHolidayName(t) || '',

  // 祝日または曜日    祝日, 日曜日...土曜日
  ff: (t, k) => PARAMS.f(t, k) || PARAMS.W(t, k),

  // 祝日
  f : (t, k) => k.getHolidayName(t) ? '祝日' : '',


  //  -----  営業日・休業日  -----

  // 曜日または休業日
  EE: (t, k) => k.isOpen(t) ? PARAMS.W(t) : '休業日',

  // 営業日または休業日
  E: (t, k) => k.isOpen(t) ? '営業日' : '休業日',

  // 休業日理由 特別休業, 祝日名, 休業日, 定休日
  e: (t, k) => k.closeCause(t),


  //  -----  時  -----

  // 24時間 0埋め    15
  HH: t => ('0' + t.getHours()).slice(-2),

  // 24時間          3, 15
  H: t => t.getHours(),

  // 12時間 0埋め    03, 15
  hh: t => ('0' + t.getHours() % 12).slice(-2),

  // 12時間          3
  h: t => t.getHours() % 12,


  //  -----  分  -----

  // 0埋め  05,  38
  II: t => ('0' + t.getMinutes()).slice(-2),

  // 0なし   5,  38
  I: t => t.getMinutes(),


  //  -----  秒  -----

  // 0埋め  07,  29
  SS: t => ('0' + t.getSeconds()).slice(-2),

  // 0なし   7,  29
  S: t => t.getSeconds(),


  // -----  ミリ秒  -----

  // 0埋め 001, 042, 567, 999
  sss: t => ('00' + t.getMilliseconds()).slice(-3),

  // 0なし   1,  42, 567, 999
  s: t => t.getMilliseconds(),

  //  -----  午前午後  -----

  // AM/PM
  AA: t => t.getHours() < 12 ? 'AM' : 'PM',

  // am/pm
  aa: t => t.getHours() < 12 ? 'am' : 'pm',

  // 午前/午後
  A: t => t.getHours() < 12 ? '午前' : '午後',


  //  ----- タイムゾーン -----

  // Z
  Z: t => {
    var tos = t.getTimezoneOffset() * -1;
    var mm = tos % 60;
    var h = (tos - mm) / 60;
    return (h < 0 ? '-' : '+') + h + ':' + ('0' + mm).slice(-2);
  },

  //  ----- 年齢 -----

  O: (t, k) => k.getAge(t),

  //  ----- 口語 -----

  K: (t, k) => k.kind(t)
};